README.md
# @ailo/money

Ailo `Money` class.

## Development

```sh
yarn
yarn start
```

## Testing

```sh
yarn lint # prettier and eslint
yarn test # unit tests
yarn test:watch # unit tests in watch mode
yarn validate # run linters, tests and build
```

## Releasing

**Note: Releasing is done manually. This project has no CI/CD configuration.**

```sh
yarn release # will automatically ask you about version bump, run tests and build, and push new version to git & npm
```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.6.1](https://github.com/ailohq/money/compare/v1.6.0...v1.6.1) (2021-05-26)

## [1.6.0](https://github.com/ailohq/money/compare/v1.5.0...v1.6.0) (2021-05-26)


### Features

* Add `roundingMode` to `multiply` and `divide` methods ([a7fa266](https://github.com/ailohq/money/commit/a7fa2662fdc1b8e0cff3ecba44bb7e665a5d4ec7))

## [1.5.0](https://github.com/ailohq/money/compare/v1.4.2...v1.5.0) (2021-03-15)


### Features

* Add `Money#divide` method ([6e91566](https://github.com/ailohq/money/commit/6e91566e7c97336728e17010217a3c68789507de))

### [1.4.2](https://github.com/ailohq/money/compare/v1.4.1...v1.4.2) (2021-03-05)

### [1.4.1](https://github.com/ailohq/money/compare/v1.4.0...v1.4.1) (2021-03-05)


### Bug Fixes

* **Money:** Do not throw error when adding/subtracting MoneyInterface ([f833380](https://github.com/ailohq/money/commit/f83338048c82a00d04e7900f2fb82cd888c00433))

## [1.4.0](https://github.com/ailohq/money/compare/v1.2.0...v1.4.0) (2021-03-03)


### Features

* added centsOnly and dollarsOnly methods (NC-107) ([7d64071](https://github.com/ailohq/money/commit/7d640713028809448bd9b2dccdba0bfe7e9b3560))


### Bug Fixes

* Allow for passing `MoneyInterface` as argument to `add` and `subtract` ([251c8ab](https://github.com/ailohq/money/commit/251c8aba4e076147f0afdbbe38dde0d7148579b1))

### [1.3.1](https://github.com/ailohq/money/compare/v1.3.0...v1.3.1) (2021-02-16)

## [1.3.0](https://github.com/ailohq/money/compare/v1.2.0...v1.3.0) (2021-02-16)


### Features

* added centsOnly and dollarsOnly methods (NC-107) ([7d64071](https://github.com/ailohq/money/commit/7d640713028809448bd9b2dccdba0bfe7e9b3560))

## [1.2.0](https://github.com/ailohq/money/compare/v1.1.0...v1.2.0) (2021-02-15)


### Features

* Add `.add` `.subtract` `.multiply` methods; remove `.diff` method ([9df41a0](https://github.com/ailohq/money/commit/9df41a0a522ce90121b863ebb9025ae98749971d))

## [1.1.0](https://github.com/ailohq/money/compare/v1.0.2...v1.1.0) (2021-02-09)


### Features

* Add `toString()` method ([98fc8b8](https://github.com/ailohq/money/commit/98fc8b8ce83a046c10b50882342cc3b524b080a1))

### [1.0.2](https://github.com/ailohq/money/compare/v1.0.1...v1.0.2) (2021-02-04)

### 1.0.1 (2021-02-02)


src/Money.test.ts
import { Money } from "./Money";

describe("Money", () => {
  describe("constructing", () => {
    it("can be constructed with and without currency", () => {
      const m1 = new Money(123);
      const m2 = new Money(123, "NZD");
      expect(m1.cents).toEqual(123);
      expect(m1.currency).toEqual("AUD");
      expect(m2.cents).toEqual(123);
      expect(m2.currency).toEqual("NZD");
    });

    it("can be constructed from an object", () => {
      const m1 = Money.from({ cents: 123 });
      const m2 = Money.from({ cents: 123, currency: "NZD" });
      expect(m1.cents).toEqual(123);
      expect(m1.currency).toEqual("AUD");
      expect(m2.cents).toEqual(123);
      expect(m2.currency).toEqual("NZD");
    });

    describe("from cents", () => {
      it("can be constructed from cents as a number", () => {
        const m1 = Money.fromCents(123);
        const m2 = Money.fromCents(123, "NZD");
        expect(m1.dollars).toEqual(1.23);
        expect(m1.currency).toEqual("AUD");
        expect(m2.dollars).toEqual(1.23);
        expect(m2.currency).toEqual("NZD");
      });

      it("can be constructed from cents as string", () => {
        const m1 = Money.fromCents("123");
        const m2 = Money.fromCents("123", "NZD");
        expect(m1.dollars).toEqual(1.23);
        expect(m1.currency).toEqual("AUD");
        expect(m2.dollars).toEqual(1.23);
        expect(m2.currency).toEqual("NZD");
      });
    });

    describe("from dollars", () => {
      it("can be constructed from dollars as a number", () => {
        const m1 = Money.fromDollars(123.45);
        const m2 = Money.fromDollars(0.12, "NZD");
        expect(m1.dollars).toEqual(123.45);
        expect(m1.currency).toEqual("AUD");
        expect(m2.dollars).toEqual(0.12);
        expect(m2.currency).toEqual("NZD");
      });

      it("can be constructed from dollars as string", () => {
        const m1 = Money.fromDollars("123.45");
        const m2 = Money.fromDollars("0.12", "NZD");
        expect(m1.dollars).toEqual(123.45);
        expect(m1.currency).toEqual("AUD");
        expect(m2.dollars).toEqual(0.12);
        expect(m2.currency).toEqual("NZD");
      });
    });

    it("if constructed from Money class, returns the same instance", () => {
      const m1 = Money.fromCents(123);
      const m2 = Money.from(m1);
      expect(m1 === m2).toEqual(true);
    });
  });

  describe("format", () => {
    it("returns formatted amount", () => {
      expect(Money.fromCents(0).format()).toEqual("$0.00");
      expect(Money.fromCents(1).format()).toEqual("$0.01");
      expect(Money.fromCents(12).format()).toEqual("$0.12");
      expect(Money.fromCents(123).format()).toEqual("$1.23");
      expect(Money.fromCents(12345).format()).toEqual("$123.45");
      expect(Money.fromCents(1234).format()).toEqual("$12.34");
      expect(Money.fromCents(123456).format()).toEqual("$1,234.56");
      expect(Money.fromCents(1234567).format()).toEqual("$12,345.67");
      expect(Money.fromCents(12345678).format()).toEqual("$123,456.78");
      expect(Money.fromCents(123456789).format()).toEqual("$1,234,567.89");
      expect(Money.fromCents(-1).format()).toEqual("-$0.01");
      expect(Money.fromCents(-12).format()).toEqual("-$0.12");
      expect(Money.fromCents(-123).format()).toEqual("-$1.23");
      expect(Money.fromCents(-1234).format()).toEqual("-$12.34");
      expect(Money.fromCents(-12345).format()).toEqual("-$123.45");
      expect(Money.fromCents(-123456).format()).toEqual("-$1,234.56");
      expect(Money.fromCents(-1234567).format()).toEqual("-$12,345.67");
      expect(Money.fromCents(-12345678).format()).toEqual("-$123,456.78");
      expect(Money.fromCents(-123456789).format()).toEqual("-$1,234,567.89");
    });

    it("can display without a symbol", () => {
      expect(Money.fromCents(123456789).format({ symbol: false })).toEqual(
        "1,234,567.89"
      );
      expect(Money.fromCents(-123456789).format({ symbol: false })).toEqual(
        "-1,234,567.89"
      );
    });

    it("can take a different thousand separator", () => {
      expect(
        Money.fromCents(123456789).format({ thousandSeparator: "" })
      ).toEqual("$1234567.89");
      expect(
        Money.fromCents(-123456789).format({ thousandSeparator: "" })
      ).toEqual("-$1234567.89");
      expect(
        Money.fromCents(123456789).format({ thousandSeparator: " " })
      ).toEqual("$1 234 567.89");
      expect(
        Money.fromCents(-123456789).format({ thousandSeparator: " " })
      ).toEqual("-$1 234 567.89");
    });

    it("can remove trailing zeroes", () => {
      expect(
        Money.fromCents(12300).format({ withTrailingZeros: false })
      ).toEqual("$123");
      expect(
        Money.fromCents(12310).format({ withTrailingZeros: false })
      ).toEqual("$123.1");
      expect(
        Money.fromCents(12311).format({ withTrailingZeros: false })
      ).toEqual("$123.11");
      expect(Money.fromCents(0).format({ withTrailingZeros: false })).toEqual(
        "$0"
      );
      expect(Money.fromCents(100).format({ withTrailingZeros: false })).toEqual(
        "$1"
      );
      expect(Money.fromCents(10).format({ withTrailingZeros: false })).toEqual(
        "$0.1"
      );
      expect(Money.fromCents(1).format({ withTrailingZeros: false })).toEqual(
        "$0.01"
      );
    });
  });

  describe("add", () => {
    it("throws an error if currencies are different", () => {
      const m1 = Money.fromCents(123, "AUD");
      const m2 = Money.fromCents(456, "NZD");

      expect(() => m1.add(m2)).toThrowErrorMatchingInlineSnapshot(
        `"Attempting to compare money of different currencies"`
      );
    });

    describe("when adding the same currency", () => {
      const m1 = Money.fromCents(100, "AUD");
      const m2 = Money.fromCents(250, "AUD");

      it("returns the difference between the two money amounts", () => {
        const fromM1ToM2 = m1.add(m2);
        expect(fromM1ToM2.dollars).toEqual(3.5);
        expect(fromM1ToM2.currency).toEqual("AUD");

        const fromM2ToM1 = m2.add(m1);
        expect(fromM2ToM1.dollars).toEqual(3.5);
        expect(fromM2ToM1.currency).toEqual("AUD");
      });
    });

    describe("when adding MoneyInterface", () => {
      const m1 = Money.fromCents(100, "AUD");
      const m2 = { cents: 250 };

      it("returns the difference between the two money amounts", () => {
        const fromM1ToM2 = m1.add(m2);
        expect(fromM1ToM2.dollars).toEqual(3.5);
        expect(fromM1ToM2.currency).toEqual("AUD");
      });
    });
  });

  describe("subtract", () => {
    it("throws an error if currencies are different", () => {
      const m1 = Money.fromCents(123, "AUD");
      const m2 = Money.fromCents(456, "NZD");

      expect(() => m1.subtract(m2)).toThrowErrorMatchingInlineSnapshot(
        `"Attempting to compare money of different currencies"`
      );
    });

    describe("when subtracting the same currency", () => {
      const m1 = Money.fromCents(100, "AUD");
      const m2 = Money.fromCents(250, "AUD");

      it("returns the difference between the two money amounts", () => {
        const diffFromM1ToM2 = m1.subtract(m2);
        expect(diffFromM1ToM2.dollars).toEqual(1.5);
        expect(diffFromM1ToM2.currency).toEqual("AUD");

        const diffFromM2ToM1 = m2.subtract(m1);
        expect(diffFromM2ToM1.dollars).toEqual(-1.5);
        expect(diffFromM2ToM1.currency).toEqual("AUD");
      });

      it("returns money with zero cents if compared to the same amount", () => {
        const diffFromM1ToItself = m1.subtract(m1);
        expect(diffFromM1ToItself.dollars).toEqual(0);
        expect(diffFromM1ToItself.currency).toEqual("AUD");
      });
    });

    describe("when subtracting MoneyInterface", () => {
      const m1 = Money.fromCents(100, "AUD");
      const m2 = { cents: 250 };

      it("returns the difference between the two money amounts", () => {
        const diffFromM1ToM2 = m1.subtract(m2);
        expect(diffFromM1ToM2.dollars).toEqual(1.5);
        expect(diffFromM1ToM2.currency).toEqual("AUD");
      });
    });
  });

  describe("multiply", () => {
    it("returns multiplied money", () => {
      expect(Money.fromCents(3).multiply(2).cents).toEqual(6);
    });

    it("by default, rounds down to the nearest cent", () => {
      expect(Money.fromCents(25).multiply(0.02).cents).toEqual(1);
      expect(Money.fromCents(50).multiply(0.02).cents).toEqual(1);
      expect(Money.fromCents(75).multiply(0.02).cents).toEqual(2);
      expect(Money.fromCents(100).multiply(0.02).cents).toEqual(2);
    });

    it("if roundingMode: 'up' is given, rounds up to the nearest cent", () => {
      expect(
        Money.fromCents(25).multiply(0.02, { roundingMode: "down" }).cents
      ).toEqual(0);
      expect(
        Money.fromCents(50).multiply(0.02, { roundingMode: "down" }).cents
      ).toEqual(1);
      expect(
        Money.fromCents(75).multiply(0.02, { roundingMode: "down" }).cents
      ).toEqual(1);
      expect(
        Money.fromCents(100).multiply(0.02, { roundingMode: "down" }).cents
      ).toEqual(2);
    });
  });

  describe("divide", () => {
    it("returns divided money", () => {
      expect(Money.fromCents(30).divide(2).cents).toEqual(15);
      expect(Money.fromCents(3).divide(2).cents).toEqual(2);
    });

    it("by default, rounds down to the nearest cent", () => {
      expect(Money.fromCents(25).divide(50).cents).toEqual(1);
      expect(Money.fromCents(50).divide(50).cents).toEqual(1);
      expect(Money.fromCents(75).divide(50).cents).toEqual(2);
      expect(Money.fromCents(100).divide(50).cents).toEqual(2);
    });

    it("if roundingMode: 'up' is given, rounds up to the nearest cent", () => {
      expect(
        Money.fromCents(25).divide(50, { roundingMode: "down" }).cents
      ).toEqual(0);
      expect(
        Money.fromCents(50).divide(50, { roundingMode: "down" }).cents
      ).toEqual(1);
      expect(
        Money.fromCents(75).divide(50, { roundingMode: "down" }).cents
      ).toEqual(1);
      expect(
        Money.fromCents(100).divide(50, { roundingMode: "down" }).cents
      ).toEqual(2);
    });
  });

  describe("absolute", () => {
    it("returns Money with positive cents if cents amount is positive", () => {
      const m1 = Money.fromCents(123, "AUD").abs();

      expect(m1.dollars).toEqual(1.23);
      expect(m1.currency).toEqual("AUD");
    });
    it("returns Money with positive cents if cents amount is negative", () => {
      const m1 = Money.fromCents(-123, "AUD").abs();

      expect(m1.dollars).toEqual(1.23);
      expect(m1.currency).toEqual("AUD");
    });
  });

  describe("isPositive", () => {
    it("returns true if cents amount is positive", () => {
      const m1 = Money.fromCents(123, "AUD");

      expect(m1.isPositive).toEqual(true);
    });
    it("returns false if cents amount is zero", () => {
      const m1 = Money.fromCents(0, "AUD");

      expect(m1.isPositive).toEqual(false);
    });
    it("returns false if cents amount is negative", () => {
      const m1 = Money.fromCents(-123, "AUD");

      expect(m1.isPositive).toEqual(false);
    });
  });

  describe("isNegative", () => {
    it("returns false if cents amount is positive", () => {
      const m1 = Money.fromCents(123, "AUD");

      expect(m1.isNegative).toEqual(false);
    });
    it("returns false if cents amount is zero", () => {
      const m1 = Money.fromCents(0, "AUD");

      expect(m1.isNegative).toEqual(false);
    });
    it("returns true if cents amount is negative", () => {
      const m1 = Money.fromCents(-123, "AUD");

      expect(m1.isNegative).toEqual(true);
    });
  });

  describe("isZero", () => {
    it("returns false if cents amount is positive", () => {
      const m1 = Money.fromCents(123, "AUD");

      expect(m1.isZero).toEqual(false);
    });
    it("returns true if cents amount is zero", () => {
      const m1 = Money.fromCents(0, "AUD");

      expect(m1.isZero).toEqual(true);
    });
    it("returns false if cents amount is negative", () => {
      const m1 = Money.fromCents(-123, "AUD");

      expect(m1.isZero).toEqual(false);
    });
  });

  describe("centsOnly", () => {
    it("returns cent component of Money", () => {
      const m1 = Money.fromCents(12345, "AUD");
      expect(m1.centsOnly).toEqual(45);
    });

    it("returns 0 for 0 cents", () => {
      const m1 = Money.fromCents(0, "AUD");
      expect(m1.centsOnly).toEqual(0);
    });

    it("returns 22 for -5422 cents", () => {
      const m1 = Money.fromCents(-5422, "AUD");
      expect(m1.centsOnly).toEqual(22);
    });
  });

  describe("dollarsOnly", () => {
    it("returns dollar component of Money", () => {
      const m1 = Money.fromCents(12345, "AUD");
      expect(m1.dollarsOnly).toEqual(123);
    });

    it("returns 0 dollars for 0 cents", () => {
      const m1 = Money.fromCents(0, "AUD");
      expect(m1.dollarsOnly).toEqual(0);
    });

    it("returns 54 for -5422 cents", () => {
      const m1 = Money.fromCents(-5422, "AUD");
      expect(m1.dollarsOnly).toEqual(54);
    });
  });

  it("`.dollars` returns dollar value of Money", () => {
    const m1 = Money.fromCents(123, "AUD");
    expect(m1.dollars).toEqual(1.23);
  });

  it("`.cents` returns cent value of Money", () => {
    const m1 = Money.fromCents(123, "AUD");
    expect(m1.cents).toEqual(123);
  });
});


src/Money.ts
import * as accounting from "accounting";
import MonetaryValue from "currency.js";

// We don't really support other currencies at the moment,
// but at least we're prepared for it ;)
export type Currency = "AUD" | "NZD";

export const DEFAULT_CURRENCY: Currency = "AUD";

export interface MoneyInterface {
  cents: number;
  currency?: Currency;
}

export class Money implements MoneyInterface {
  readonly currency: Currency;

  private readonly value: MonetaryValue;

  constructor(
    cents: number | string | MonetaryValue,
    currency: Currency = DEFAULT_CURRENCY
  ) {
    this.value = MonetaryValue(cents, { fromCents: true });
    this.currency = currency;
  }

  static from(obj: MoneyInterface | Money): Money {
    if (obj instanceof Money) {
      return obj;
    }

    return new Money(obj.cents, obj.currency);
  }

  static fromCents(
    cents: number | string,
    currency: Currency = DEFAULT_CURRENCY
  ): Money {
    const value = MonetaryValue(cents, { fromCents: true });
    return new Money(value, currency);
  }

  static fromDollars(
    dollars: number | string,
    currency: Currency = DEFAULT_CURRENCY
  ): Money {
    const value = MonetaryValue(dollars);
    return new Money(value, currency);
  }

  static zero(currency: Currency = DEFAULT_CURRENCY): Money {
    return new Money(0, currency);
  }

  /**
   * Returns float value of the dollars amount,
   * e.g. `123.45` for `{ cents: 12345 }`.
   */
  get dollars(): number {
    return this.value.value;
  }

  /**
   * Returns only a whole dollar component. 123 for {cents: 12345}
   */
  get dollarsOnly(): number {
    return Math.abs(this.value.dollars());
  }

  /**
   * Returns int value of the cents amount,
   * e.g. `12345` for `{ cents: 12345 }`.
   */
  get cents(): number {
    return this.value.intValue;
  }

  /**
   * Returns only cents component. 45 for {cents: 12345}
   */
  get centsOnly(): number {
    return Math.abs(this.value.cents());
  }

  get isPositive(): boolean {
    return this.value.value > 0;
  }

  get isNegative(): boolean {
    return this.value.value < 0;
  }

  get isZero(): boolean {
    return this.value.value === 0;
  }

  add(addend: MoneyInterface): Money {
    if (addend.currency && this.currency !== addend.currency) {
      throw new TypeError(
        "Attempting to compare money of different currencies"
      );
    }

    return new Money(Money.from(addend).value.add(this.value), this.currency);
  }

  subtract(subtrahend: MoneyInterface): Money {
    if (subtrahend.currency && this.currency !== subtrahend.currency) {
      throw new TypeError(
        "Attempting to compare money of different currencies"
      );
    }

    return new Money(
      Money.from(subtrahend).value.subtract(this.value),
      this.currency
    );
  }

  multiply(
    multiplier: number,
    {
      roundingMode = "up",
    }: {
      /**
       * @default "up"
       */
      roundingMode?: "down" | "up";
    } = {}
  ): Money {
    if (roundingMode === "down") {
      return Money.fromCents(
        Math.floor(this.cents * multiplier),
        this.currency
      );
    }

    return new Money(this.value.multiply(multiplier), this.currency);
  }

  divide(
    multiplier: number,
    {
      roundingMode = "up",
    }: {
      /**
       * @default "up"
       */
      roundingMode?: "down" | "up";
    } = {}
  ): Money {
    if (roundingMode === "down") {
      return Money.fromCents(
        Math.floor(this.cents / multiplier),
        this.currency
      );
    }

    return new Money(this.value.divide(multiplier), this.currency);
  }

  abs(): Money {
    if (this.isPositive || this.isZero) {
      return this;
    }

    return new Money(this.value.multiply(-1), this.currency);
  }

  format({
    symbol = true,
    thousandSeparator = ",",
    withTrailingZeros = true,
  }: {
    /**
     * @default true
     */
    symbol?: boolean;
    /**
     * @default ","
     */
    thousandSeparator?: "," | "" | " ";
    /**
     * @default true
     */
    withTrailingZeros?: boolean;
  } = {}): string {
    const result = accounting.formatMoney(this.dollars, {
      symbol: symbol ? "$" : "",
      precision: 2,
      thousand: thousandSeparator,
      format: {
        pos: "%s%v",
        neg: "-%s%v",
        zero: "%s%v",
      },
    });

    if (!withTrailingZeros && result.includes(".")) {
      if (result.slice(-2) === "00") {
        return result.slice(0, -3);
      }
      if (result.slice(-1) === "0") {
        return result.slice(0, -1);
      }
    }

    return result;
  }

  toString(): string {
    return this.format();
  }
}
